### ISSUE TITLE
Connecting Box.com gives error

### ISSUE THREAD

**Issue by bratislav** [spring-guides/tut-spring-boot-oauth2#21]
Hi,

I'm trying to configure github example to additionally connect to my Box.com but without luck. Here's what I've set in application.yml :

box:
&nbsp;&nbsp;client:
&nbsp;&nbsp;&nbsp;&nbsp;clientId: xxx
&nbsp;&nbsp;&nbsp;&nbsp;clientSecret: yyy
&nbsp;&nbsp;&nbsp;&nbsp;accessTokenUri: https://www.box.com/api/oauth2/token
&nbsp;&nbsp;&nbsp;&nbsp;userAuthorizationUri: https://www.box.com/api/oauth2/authorize
&nbsp;&nbsp;&nbsp;&nbsp;clientAuthenticationScheme: form
&nbsp;&nbsp;resource:
&nbsp;&nbsp;&nbsp;&nbsp;userInfoUri: https://api.box.com/2.0/users/me

After updating java code as follows :

```java
@Bean
@ConfigurationProperties("box")
ClientResources box() {
	return new ClientResources();
}
...

private Filter ssoFilter() {
	CompositeFilter filter = new CompositeFilter();
	List<Filter> filters = new ArrayList<>();
	filters.add(ssoFilter(facebook(), "/login/facebook"));
	filters.add(ssoFilter(github(), "/login/github"));
	filters.add(ssoFilter(box(), "/login/box"));
	filters.add(ssoFilter(google(), "/login/google"));
	filter.setFilters(filters);
	return filter;
}
```

...and updating HTML page with the additional link, I'm getting this error :

```
Tue May 31 15:11:55 CEST 2016
There was an unexpected error (type=Unauthorized, status=401).
Authentication Failed: Could not obtain user details from token
```

Any clues? Thanks in advance.

**Comment by dsyer** [comment]
I never used box.com as an identity provider before. Do they document their authentication process? Probably you just failed to follow it in some way. Spring OAuth2 follows the spec fairly closely, but you can sometimes fall in the cracks, and identity management is not part of the core bearer token spec, so different providers implement it in different ways.

**Comment by bratislav** [comment]
Yes, they do, it's fully documented here : https://docs.box.com/docs/oauth-20

I've read it and tried to fire requests manually over Postman and of course everything works...but no luck over Spring. Just spotted an error inside the log :

2016-05-31 14:36:03.831 DEBUG 55892 --- [http-nio-8080-exec-9] o.s.s.oauth2.client.OAuth2RestTemplate   : GET request for "https://api.box.com/2.0/users/me" resulted in 400 (Bad Request); invoking error handler

How can I try to pinpoint the actual problem, is there any diagnostic I can run? Tried to catch traffic over Fiddler didn't reveal much. If you can at least direct me in the right way, I'd be much appreciated.

dsyer please help :)

**Comment by dsyer** [comment]
Providers are not very open to telling you why your request is bad. Some of them even just send a 404. But if you have an example of a successful request to /users/me then you can compare it to the one that fails and sport the difference. Probably a header spelling difference or something stupid.

**Comment by bratislav** [comment]
Hm, I undestand, put when I look at the Fiddler caputred traffic I'm not seeing any calls to the /users/me. Here's the JPG attached
![401](https://cloud.githubusercontent.com/assets/4982653/15677496/cb92657e-274a-11e6-9a7a-5d0018c0da72.jpg)

How can I customize header, what calss shoudl I add/extend?

**Comment by dsyer** [comment]
Isn't that a 401 from Spring Security then? Did you forget to `permitAll()` on /login/box?

**Comment by bratislav** [comment]
Ouch...ahm, how can I do that? Just started playing with Spring, still learning how that works.

Is this line suppose to do that :

.antMatchers("/", "/login**", "/webjars/**").permitAll() 

This is taken from your code, didn't touch any of that.

**Comment by dsyer** [comment]
I just tried it with my own box app. They expect the authorization header to be "Bearer ..." (not "bearer ...") which is the same as google IIRC. So if you have google working, then the same trick should work for this.

**Comment by dsyer** [comment]
TL;DR you need a `OAuth2RequestAuthenticator` injected into the user info rest template. So create a `UserInfoRestTemplateCustomiizer` and make it a `@Bean`. In the `customize()` method inject your authenticator, and pull the token value out of the client context it gives you.

**Comment by bratislav** [comment]
Thanks Dave. To wrap up, I need to create two new classes:

- one that extends DefaultOAuth2RequestAuthenticator 
- the second one that implements UserInfoRestTemplateCustomizer

One they're done how can I hook them into the github example?
Can I use snippets from this link https://github.com/spring-projects/spring-security-oauth/issues/457 to start with?

**Comment by bratislav** [comment]
Here are the files I came up with so far :

`package com.example;

import org.springframework.context.annotation.Bean;
import org.springframework.security.oauth2.client.OAuth2RestTemplate;
import org.springframework.boot.autoconfigure.security.oauth2.resource.UserInfoRestTemplateCustomizer;

public class BoxTemplateCustomizer implements UserInfoRestTemplateCustomizer  {

    @Override
    @Bean
    public void customize(OAuth2RestTemplate template) {
        template.setAuthenticator(new BoxRequestAuthenticator());
    }
}`

`package com.example;

import org.springframework.security.oauth2.client.DefaultOAuth2RequestAuthenticator;
import org.springframework.security.oauth2.client.resource.OAuth2ProtectedResourceDetails;
import org.springframework.security.oauth2.client.OAuth2ClientContext;
import org.springframework.http.client.ClientHttpRequest;
import org.springframework.security.oauth2.common.OAuth2AccessToken;

public class BoxRequestAuthenticator extends DefaultOAuth2RequestAuthenticator  {

    @Override
    public void authenticate(OAuth2ProtectedResourceDetails resource, OAuth2ClientContext clientContext, ClientHttpRequest request) {
        OAuth2AccessToken accessToken = clientContext.getAccessToken();
        String tokenType = OAuth2AccessToken.BEARER_TYPE;
        request.getHeaders().set("Authorization", String.format("%s %s", tokenType, accessToken.getValue()));
    }
    
}`

Are these ok? How to hook them into the project?

**Comment by dsyer** [comment]
Take the `@Bean` annotation off the service method and add a `@Component` annotation at the type level, might be a good start. Then go and read the Spring User guide, or a few basic getting started guides on spring.io.
